psi4/src/psi4/dlpno/sparse.cc

```diff
@@ -0,0 +1,219 @@
+/*
+ * @BEGIN LICENSE
+ *
+ * Psi4: an open-source quantum chemistry software package
+ *
+ * Copyright (c) 2007-2021 The Psi4 Developers.
+ *
+ * The copyrights for code used from other parties are included in
+ * the corresponding files.
+ *
+ * This file is part of Psi4.
+ *
+ * Psi4 is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * Psi4 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along
+ * with Psi4; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * @END LICENSE
+ */
+
+#include "sparse.h"
+
+#include <tuple>
+#include <algorithm>
+
+namespace psi {
+
+/* Args: sorted lists l1 and l2
+ * Returns: sorted union of l1 and l2
+ */
+std::vector<int> merge_lists(const std::vector<int> &l1, const std::vector<int> &l2) {
+
+    std::vector<int> l12;
+
+    int i1 = 0, i2 = 0;
+    while(i1 < l1.size() || i2 < l2.size()) {
+        if(i1 == l1.size()) {
+            l12.push_back(l2[i2]);
+            ++i2;
+        } else if(i2 == l2.size()) {
+            l12.push_back(l1[i1]);
+            ++i1;
+        } else if(l1[i1] == l2[i2]) {
+            l12.push_back(l1[i1]);
+            ++i1;
+            ++i2;
+        } else if(l1[i1] < l2[i2]) {
+            l12.push_back(l1[i1]);
+            ++i1;
+        } else {
+            l12.push_back(l2[i2]);
+            ++i2;
+        }
+    }
+
+    return l12;
+
+}
+
+/* Args: sorted list of y, sparse map from A to y (assume sorted)
+ * Returns: sorted list yA
+ *
+ * For all a, every value in A_to_y[a] is included in yA if at least one is present in y
+ */
+std::vector<int> contract_lists(const std::vector<int> &y, const std::vector<std::vector<int>> &A_to_y) {
+
+    // TODO: runtime is proportional to A_to_y size (system size, O(N))
+    // could maybe reduce to &y size (domain size, O(1)), probably doesn't matter
+    std::vector<int> yA;
+
+    for(int a = 0, y_ind = 0; a < A_to_y.size(); ++a) {
+
+        bool is_a = false;
+        for(auto y_val : A_to_y[a]) {
+            if (y_ind < y.size() && y[y_ind] == y_val) {
```

Comment:
Is it not more efficient to use `std::find(y.begin(), y.end(), y_val)`?

Reply:
When `contract_lists` is called, the list `y` may be very long, but each list `A_to_y[a]` will be small, so it's more efficient to search through `A_to_y[a]` as currently written.

Also in the bigger picture, this part of the code is so fast compared to computing integrals and PNOs, that it doesn't really matter either way.